File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode head2 = head;
        //fast and slow (1 step vs 2 steps)
        //if the fast and slow pointer ever meets at the same node, return true
        if (head.next == null){
            return false;
        }
        while ((head != null) && (head2 != null)){
            head = head.next;
            head2 = head2.next.next;
            if (head == head2){
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # reorder the list so that it follows a jumping order
        # do fast and slow pointer approach
        # reverse the second half (slow to the end)
        slow = head
        fast = head
        while fast and fast.next:
            print(fast.val)
            slow = slow.next
            fast = fast.next.next
        
        # slow is currently at the midpoint
        # 1 2 3 4 null
        print(slow.val)
        curr = slow
        prev = None
        while curr:
            print(f"1 {curr.val}")
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        
        # prev points to the second half head

        # repoint head and slow
        # 1 2 3
        # 4 5
        start = head
        while prev:
            print(f"2 {start.val} {prev.val}")
            start_nxt = start.next
            slow_nxt = prev.next
            start.next = prev
            if start_nxt != prev:
                prev.next = start_nxt
            start = start_nxt
            prev = slow_nxt
        
        # we need to check to make sure start_nxt != prev
        

        
